enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java

```diff
@@ -117,24 +114,42 @@ public synchronized void notifyCommitted( long commitIndex )
     private void submitApplyJob( long lastToApply )
     {
         applier.submit( ( status ) -> () -> {
-            try ( RaftLogCursor cursor = raftLog.getEntryCursor( lastApplied + 1 ) )
+            try ( LogEntrySupplier cacheLogEntrySupplier = new CacheApplier();
+                    LogEntrySupplier cursorLogEntrySupplier = new CursorApplier() )
             {
-                lastApplyingStorage.persistStoreData( lastToApply );
-
-                while ( cursor.index() < lastToApply && cursor.next() )
+                LogEntrySupplier currentLogEntrySupplier = cacheLogEntrySupplier;
+                for ( long i = lastApplied + 1; i < lastToApply; i++ )
                 {
-                    if ( cursor.get().content() instanceof DistributedOperation )
+                    lastApplyingStorage.persistStoreData( lastToApply );
+                    RaftLogEntry raftLogEntry = currentLogEntrySupplier.get( i );
+                    if ( raftLogEntry == null )
                     {
-                        DistributedOperation distributedOperation = (DistributedOperation) cursor.get().content();
-
-                        progressTracker.trackReplication( distributedOperation );
-                        handleOperation( cursor.index(), distributedOperation );
+                        if ( currentLogEntrySupplier == cursorLogEntrySupplier )
+                        {
+                            throw new IllegalStateException(
+                                    format( "Index %d not found in the raft log. lastToApply was: %d, append index " +
+                                            "is %d", i, lastToApply, raftLog.appendIndex() ) );
+                        }
+                        currentLogEntrySupplier = cursorLogEntrySupplier;
+                        raftLogEntry = currentLogEntrySupplier.get( i );
+                    }
 
-                        maybeFlush();
+                    if ( raftLogEntry == null )
+                    {
+                        String failureMessage = format(
+                                "Log entry at index %d is not present in the log. Please perform recovery by " +
+                                        "restarting this database instance. Indexes to be applied are %d up to %d, " +
+                                        "append index is %d, prev index is %d.",
+                                i, lastApplied + 1, lastToApply, raftLog.appendIndex(), raftLog.prevIndex() );
+                        log.error( failureMessage );
+                        throw new IllegalStateException( failureMessage );
```

Comment:
I think having a `DelegateRaftLog` class that will try to use the cache first or call the real raftlog implementation is a better way to abstract the cache away while reading